src/Features/Core/Portable/Diagnostics/CodeAnalysisDiagnosticAnalyzerService.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+            _diagnosticsRefresher = diagnosticsRefresher;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+            => new CodeAnalysisDiagnosticAnalyzerService(_diagnosticAnalyzerService, _diagnosticsRefresher, workspaceServices.Workspace);
+
+        private sealed class CodeAnalysisDiagnosticAnalyzerService : ICodeAnalysisDiagnosticAnalyzerService
+        {
+            private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+            private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+            private readonly Workspace _workspace;
+            private readonly ConcurrentSet<ProjectId> _analyzedProjectIds = new();
+
+            public CodeAnalysisDiagnosticAnalyzerService(
+                IDiagnosticAnalyzerService diagnosticAnalyzerService,
+                IDiagnosticsRefresher diagnosticsRefresher,
+                Workspace workspace)
+            {
+                _diagnosticAnalyzerService = diagnosticAnalyzerService;
+                _diagnosticsRefresher = diagnosticsRefresher;
+                _workspace = workspace;
+
+                _workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+            {
+                switch (e.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                        _analyzedProjectIds.Clear();
+                        break;
+                }
+            }
+
+            public bool HasProjectBeenAnalyzed(ProjectId projectId) => _analyzedProjectIds.Contains(projectId);
+
+            public async Task RunAnalysisAsync(Solution solution, Action<Project> onProjectAnalyzed, ProjectId? projectId, CancellationToken cancellationToken)
+            {
+                if (projectId != null)
+                {
+                    var project = solution.GetProject(projectId);
+                    if (project != null)
+                    {
+                        await AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    var tasks = new Task[solution.ProjectIds.Count];
+                    var index = 0;
+                    foreach (var project in solution.Projects)
+                    {
+                        tasks[index++] = Task.Run(() => AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken), cancellationToken);
+                    }
+
+                    await Task.WhenAll(tasks).ConfigureAwait(false);
```

Comment:
```suggestion
                    using var _ = ArrayBuilder.GetInstance<Task>(solution.Projects.Count), out var tasks);
                    foreach (var project in solution.Projects)
                        tasks.Add(Task.Run(() => AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken), cancellationToken));

                    await Task.WhenAll(tasks).ConfigureAwait(false);
```

Reply:
note: mention that because thsi is user kicked off (not ambient), it's reasonable to be concurrent and use all the machine resources.